#  GENETIC ALGORITHM

# Optimization of advertising investment

The following problem tries to solve the best combination of investment in advertising through Genetic Algorithms.

These algorithms make a population of individuals evolve, subjecting them to random actions similar to biological evolution, in order to find optimization solutions.

https://es.wikipedia.org/wiki/Algoritmo_gen%C3%A9tico#:~:text=Los%20algoritmos%20gen%C3%A9ticos%20(AG)%20funcionan,la%20cadena%20son%20llamados%20genes.

**IMPORT LIBRARIES**

In [1]:
import random
import numpy as np
from random import randrange
from random import randint, uniform

**MUTATION FUNCTION**

In [2]:
def mutateIndividual(ind):
    """Get mutated or non-mutated value
    
    Parameters:
    -----------
    ind: lst
        contains a list with the values of the child pattern
    
    Return:
    -------
    ind: lst
        contains a list with the mutated or non-mutated values of the child pattern
    
    """
    for i in range(len(ind)): # select one chromosome
        if random.random() < 0.15: # random probability of mutation
            if ind[i] == 0:
                ind[i] = 1
            else:
                ind[i] = 0
    return ind

**GENETIC ALGORITHM FUCTION**

In [3]:
def BinaryTournamentAlgorithm(generations, costs, profit, limit):
    """Get the best investment combination through genetic algorithms
    
    Parameters:
    -----------
    
    generations: int 
        number of iterations the evaluation is performed
    
    costs: lst
        list of costs for each advertising medium
    
    profit: lst
        list of earnings of each advertising medium
    
    limit: int
        maximum investment value
        
    Return: 
    -------
    best_supports: lst
        optimized list of advertising media
        
    """
    
    P = []    # random genotype initialization
    for j in range(len(costs)):
        p = []
        for v in range(len(costs)):
            p.append(randrange(2))
        P.append(p)
    
    for q in range(generations):   # number of iterations the evaluation is performed 
        fenotipos = []
        for c in P:
            fenotipos.append(sum(c*profit))

        parents = []
        parents_index = []
        for s in range(2):    # parent selection by binary tournament
            x = random.sample(range(0,len(costs)),3)
            selected = []
            for n in x:
                selected.append(fenotipos[n])
            parents_index.append(x[np.argmax(np.array(selected))])
            parents.append(P[x[np.argmax(np.array(selected))]])

        x = randrange(1,20)    # crossing
        child1 = parents[0][0:x]+parents[1][x:]
        child2 = parents[1][0:x]+parents[0][x:]

        child1 = mutateIndividual(child1)    # mutation
        child2 = mutateIndividual(child2)

        similar_parent0h1 = 0    # replacement
        similar_parent1h1 = 0
        similar_parent0h2 = 0
        similar_parent1h2 = 0

        for i in range(len(costs)):
            if parents[0][i] == child1[i]:
                similar_parent0h1 +=1
            if parents[1][i] == child1[i]:
                similar_parent1h1 +=1
            if parents[0][i] == child2[i]:
                similar_parent0h2 +=1
            if parents[1][i] == child2[i]:
                similar_parent1h2 +=1

        if similar_parent0h1 == similar_parent1h1:
            if similar_parent0h2 == similar_parent1h2:
                P[parents_index[0]] = child1
                P[parents_index[1]] = child2
            elif similar_parent0h2 > similar_parent1h2:
                P[parents_index[0]] = child2
                P[parents_index[1]] = child1
        elif similar_parent0h2 == similar_parent1h2:
            if similar_parent0h1 == similar_parent1h1:
                P[parents_index[1]] = child1
                P[parents_index[0]] = child2
            elif similar_parent0h1 > similar_parent1h1:
                P[parents_index[1]] = child2
                P[parents_index[0]] = child1
        elif similar_parent0h1 > similar_parent1h1:
            P[parents_index[0]] = child1
            P[parents_index[1]] = child2
        elif similar_parent0h1 < similar_parent1h1:
            P[parents_index[1]] = child1
            P[parents_index[0]] = child2
        elif similar_parent0h2 > similar_parent1h2:
            P[parents_index[1]] = child1
            P[parents_index[0]] = child2
        elif similar_parent0h2 < similar_parent1h2:
            P[parents_index[0]] = child1
            P[parents_index[1]] = child2
            
        net_profit = []
        gross_profit = []
        cost = []
        final_P = []
        
        for w in range(len(costs)):
            if (sum(P[w]*costs))<=limit:
                gross_profit.append((sum(P[w]*profit)))
                cost.append((sum(P[w]*costs)))                                       
                net_profit.append((sum(P[w]*profit)-(sum(P[w]*costs))))
                final_P.append(P[w])
            else:
                pass
    try:
        print('Investment: ', cost[net_profit.index(max(net_profit))])
        print('Gross Profit: ', gross_profit[net_profit.index(max(net_profit))])
        print('Net profit: ', net_profit[net_profit.index(max(net_profit))])
        print('\nThe best advertising combination: ')
        best_combination = final_P[net_profit.index(max(net_profit))]
    except:
        best_combination = None
        print('No combination has been found. Increase the number of generations')

    best_supports = []
    for x, y in zip(supports, best_combination):
        best_supports.extend([x] * y)
    
    return best_supports

**DATA**

The data that will be taken in the model are declared below:

- The **limit** refers to the maximum investment considered

- The **generations** are the iterations that the algorithm will perform for its evolution

- The **supports** are the descriptive label of each advertising support

- The **costs** are the cost of each advertising action

- The **benefit** is the return on investment of each advertising action

(These data are fictitious)

In [4]:
limit = 5000
generations = 500
supports = ['googleads', 'fb', 'ig', 'radio', 'billboards', 'periódico', 'newspaper', 'twitter', 'flyers', 'stands', 'roll up', 'high reputation influencer', 'raffles', 'tv high hours', 'advertising on the subway', 'advertising at bus stops', 'medium reputation influencer', 'low reputation influencer', 'tv low hours', 'mobile applications']
costs = np.array([300, 60, 300, 1000, 400, 500, 500, 1500, 200, 900, 1300, 300, 700, 1500, 800, 500, 1000, 300, 800, 200])
profit = np.array([600, 200, 500, 1200, 700, 900, 900, 1600, 500, 2000, 1500, 500, 1000, 1900, 1000, 800, 1500, 600, 900, 450])

**RESULTS**

In [5]:
BinaryTournamentAlgorithm(generations, costs, profit, limit)

Investment:  4460
Gross Profit:  7350
Net profit:  2890

The best advertising combination: 


['fb',
 'radio',
 'periódico',
 'stands',
 'raffles',
 'advertising on the subway',
 'low reputation influencer',
 'mobile applications']

As a conclusion, the investment required for this optimized combination is detailed. Being 4,460€ and obtaining a net profit of 2,890€. 

This strategy recovers the total capital invested and also achieves a 64% profit.